runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java

```diff
@@ -479,7 +479,32 @@ public void processWatermark1(Watermark mark) throws Exception {
 
   @Override
   public void processWatermark2(Watermark mark) throws Exception {
-    // ignore watermarks from the side-input input
+    if (mark.getTimestamp() == BoundedWindow.TIMESTAMP_MAX_VALUE.getMillis()) {
+      // this means we will never see any more side input
+      pushbackDoFnRunner.startBundle();
+
+      BagState<WindowedValue<InputT>> pushedBack =
+          pushbackStateInternals.state(StateNamespaces.global(), pushedBackTag);
+
+      Iterable<WindowedValue<InputT>> pushedBackContents = pushedBack.read();
+      if (pushedBackContents != null) {
+        for (WindowedValue<InputT> elem : pushedBackContents) {
+
+          // we need to set the correct key in case the operator is
+          // a (keyed) window operator
+          setKeyContextElement1(new StreamRecord<>(elem));
+
+          pushbackDoFnRunner.processElement(elem);
+        }
+      }
+
+      setPushedBackWatermark(BoundedWindow.TIMESTAMP_MAX_VALUE.getMillis());
+
+      pushbackDoFnRunner.finishBundle();
+
+      // maybe output a new watermark
```

Comment:
I think roughly yes, though not all in this method. This method just does:
 - notice that we get `+Inf` watermark on side input -> we will never again see any side input
 - process pushed back values, even though their side input might be empty (some failing tests actually verified that this was the case)
 - update the output watermark. It might still be held back by the watermark hold in case this operator is a `WindowDoFnOperator`. That's why we go through that logic in `processWatermark1()` instead of simply forwarding the `+Inf` watermark.